Modules/Core/Common/include/itkMath.h

```diff
@@ -762,6 +762,54 @@ GreatestPrimeFactor(unsigned long n);
 ITKCommon_EXPORT unsigned long long
 GreatestPrimeFactor(unsigned long long n);
 
+// C++11 does not guarantee that assert can be used in constexpr
+// functions. This is a work-around for GCC 4.8, 4.9. Originating
+// from Andrzej's C++ blog:
+//  https://akrzemi1.wordpress.com/2017/05/18/asserts-in-constexpr-functions/
+#if defined NDEBUG
+#  define ITK_X_ASSERT(CHECK) void(0)
+#else
+#  define ITK_X_ASSERT(CHECK) ((CHECK) ? void(0) : [] { assert(!#CHECK); }())
+#endif
+
+namespace Experimental
+{
+/**  Returns `a * b`. Numeric overflow triggers a compilation error in
+ * "constexpr context" and a debug assert failure at run-time.
+ */
+constexpr std::uintmax_t
+UnsignedProduct(const std::uintmax_t a, const std::uintmax_t b) ITK_NOEXCEPT
+{
+  // Note that numeric overflow is not "undefined behavior", for unsigned numbers.
+  // This function checks if the result of a*b is mathematically correct.
+  return (a == 0) || (b == 0) || ((((a * b) / a) == b) && (((a * b) / b) == a))
+           ? (a * b)
+           : (ITK_X_ASSERT(!"UnsignedProduct overflow!"), 0);
+}
+
+
+/** Calculates base ^ exponent. Numeric overflow triggers a compilation error in
+ * "constexpr context" and a debug assert failure at run-time. Otherwise equivalent to
+ * C++11 `static_cast<std::uintmax_t>(std::pow(base, exponent))`
+ *
+ * \note `UnsignedPower(0, 0)` is not supported, as zero to the power of zero does
+ * have no agreed-upon value: https://en.wikipedia.org/wiki/Zero_to_the_power_of_zero
+ */
+constexpr std::uintmax_t
+UnsignedPower(const std::uintmax_t base, const std::uintmax_t exponent) ITK_NOEXCEPT
```

Comment:
You should template this method over the output type. I don't think having exponent be a larger type than unsigned byte is useful.

Reply:
@blowekamp 
> I don't think there is a need to place these little functions into an Experimental namespace

My main reason for putting them in an `Experimental` namespace is because I was not 100% sure about the names of the new functions. I thought about `CalculateUnsignedPower`,  `CalculatePowerOfUnsigned`, etc., and `CalculateUnsignedProduct` or `MultiplyUnsigned`...

But I think the current function names are also fine. 😃 So if no one objects, I'll adjust the PR, so that they are directly placed in the `itk::Math` namespace.

Reply:
@blowekamp 
> You should template this method over the output type

The result can easily become very large, that 's why I thought `std::uintmax_t` would be appropriate as a return type. (Note that on most platforms, `uintmax_t` and `size_t` have the same size anyway.)

> I don't think having exponent be a larger type than unsigned byte is useful.

A common use case would be to have an `ImageDimension` as exponent, which is often defined as 32-bit or 64-bit unsigned. I would not want to require the user to explicitly cast their argument from 32/64-bit to byte, when using the value as exponent. Would you?

Can you show me how you would want the function to look like?

Reply:
> The result can easily become very large, that 's why I thought std::uintmax_t would be appropriate as a return type. (Note that on most platforms, uintmax_t and size_t have the same size anyway.)

This would not be true on 32-bit platforms.

>A common use case would be to have an ImageDimension as exponent, which is often defined as 32-bit or 64-bit unsigned. I would not want to require the user to explicitly cast their argument from 32/64-bit to byte, when using the value as exponent. Would you?

I agree avoiding casting is good, and that is logical for the exponent type. The same reasoning can be made for having a templated output type. 😄 

```
template <typename TValueType=std::uintmax_t>
constexpr TValueType
UnsignedPower(const TValueType base, const std::uintmax_t exponent)
```

Reply:
Thanks for your suggestion @blowekamp Unfortunately I haven't got it working yet. When I modified the signature as you suggested, the test `UnsignedPower(2, 3)` deduced `TValueType = int`.  It completely ignored the default, `TValueType=std::uintmax_t`, which I would like to have! I'll have another look...

Reply:
Ahh, templates are a complicated thing. It should have been obvious that the temple parameters will automatically be deduced, and the default parameter for the function should not make much sense.

Reply:
@blowekamp I'm preparing a force-push, which allows specifying the return type of `UnsignedPower`. It will still be `uintmax_t` by default, though. Hope that's OK to you. Please wait  😄

Reply:
@blowekamp My last force-push does include compile-time tests, that the return type of a function call `UnsignedPower<T>(1, 1)` is `T`. Please check: 

https://github.com/InsightSoftwareConsortium/ITK/blob/996f95938ed19da6e9166ee7a2a784b64db2b38f/Modules/Core/Common/test/itkMathTest.cxx#L50